File: client/app/src/main/java/com/ezntek/beanmaths/components/Component.java
package com.ezntek.beanmaths.components;

import static com.ezntek.beanmaths.components.ComponentState.*;

public abstract class Component {
    public ComponentState cmpState;

    public static Component makeComponent() {
        class DummyComponent extends Component {
            public void render() {
            }

            public void update(long gtState) {
            }
        }

        return new DummyComponent();
    }

    public abstract void render();

    public abstract void update(long gtState);

    public boolean shouldRender() {
        return !(this.cmpState == DISABLED || this.cmpState == NODRAW);
    }

    public boolean shouldUpdate() {
        return !(this.cmpState == DISABLED || this.cmpState == NOUPDATE);
    }
}

File: client/src/components/component.cpp
#include "./component.hpp"

namespace components {

bool Component::shouldRender() const {
    return !(this->componentState == ComponentState::NORENDER ||
             this->componentState == ComponentState::DISABLED);
}

bool Component::shouldUpdate() const {
    return !(this->componentState == ComponentState::NOUPDATE ||
             this->componentState == ComponentState::DISABLED);
}

} // namespace components